src/text/raw_text_reader.rs

```diff
@@ -480,170 +99,106 @@ impl<T: ToIonDataSource> IonReader for RawTextReader<T> {
     }
 
     fn next(&mut self) -> IonResult<RawStreamItem> {
-        // Parse the next value from the stream, storing it in `self.current_value`.
-        self.load_next_value()?;
-
-        // If we're positioned on an IVM, return the (major, minor) version tuple
-        if let Some((major, minor)) = self.current_ivm {
-            return Ok(RawStreamItem::VersionMarker(major, minor));
-        }
+        self.feed_if_needed()?;
```

Comment:
Rather than using the (reasonable!) heuristic you've implemented on each call to `next()`, I propose:
1. Doing an initial read at reader creation time. Our [builder APIs return an `IonResult<ReaderType>`](https://github.com/amzn/ion-rust/blob/94c83bf336f0c3902ac81b594f5d38774c097782/src/reader.rs#L37) to accommodate I/O during construction. (In that particular example, we peek at the first four bytes of the file to detect whether we're dealing with text or binary.)
2. Only reading from the source when we encounter an `Incomplete`.

That way most calls to `next()` will be free in terms of I/O checking.

Reply:
I think we need a way for the user to flag the non-blocking reader as 'done' in terms of data being provided. The reason I started reading ahead of consuming functions was to reduce the chances of the parser returning back values and advancing when partial values were present.

For example, in the unit tests that used the builder APIs, the chain of first 4 bytes followed by the rest of the buffer, resulted in a first `read` of 4 bytes. For text encoded timestamps, that lined up perfectly with the 4 digit year, which would parse as an integer. Then the next read would generate an error. Those boundaries seem to be possible anywhere, and there seems to be ample opportunities for partial values to be parsed as valid values.

What do you think of having the user signal when data is complete, and if that signal has not been made and a parse would consume the remainder of the buffer, an incomplete error is returned and the data is left as is?

With that I think we can rely on the `Incomplete` to signal when to read.

Reply:
Oh! Thank you for explaining the problem. Take a look at the old `RawTextReader`'s [`parse_value_at_eof()` method](https://github.com/amzn/ion-rust/blob/main/src/text/raw_text_reader.rs#L376). What we need (and I think this is similar to what you described) is a way for the blocking reader to say:
```
if self.reader.next() == Incomplete and self.input().is_eof() {
  return self.reader.parse_value_at_eof();
}
```
What I don't know (and we don't necessarily need to solve this here/now) is how to do this in a way that's format agnostic. The binary reader doesn't have this problem; it can always tell where the boundaries of a value are.

We may end up needing to add some methods to the `RawReader` trait.